shivyc/il_cmds/math.py

```diff
@@ -211,3 +211,33 @@ class Mod(_DivMod):
     """
 
     return_reg = spots.RDX
+
+
+class Neg(ILCommand):
+    """Negates given IL value.
+
+    No type promotion is done here.
+
+    """
+
+    def __init__(self, output, arg):  # noqa D102
+        self.output = output
+        self.arg = arg
+
+    def inputs(self):  # noqa D102
+        return [self.arg]
+
+    def outputs(self):  # noqa D102
+        return [self.output]
+
+    def rel_spot_pref(self):  # noqa D102
+        return {self.output: [self.arg]}
+
+    def make_asm(self, spotmap, home_spots, get_reg, asm_code): # noqa D102
```

Comment:
Thanks for figuring this out on your own. Your implementation outputs correct assembly AFAIK, and there's a couple quick changes you can make to improve the efficiency of the emitted code :)

Some background on how the register allocation work in ShivyC: the register allocator looks at the inputs, outputs, rel_spot_pref, etc. of each of the IL commands to decide how to best allocate registers to all of the variables (check out the il_cmds/base.py file for explicit documentation on what each of these mean). In this case, the `{self.output: [self.arg]}` spot preference tells the register allocator that if possible, we want the `self.output` IL value to be assigned to the same spot (i.e. register) as `self.arg`. This is helpful because if the register allocator is able to accomplish this, then we don't have to emit a `mov` command since the argument is already in the output spot. It turns out that about 70% of the time, the register allocator is able to meet these preferences. To take advantage of this, then, you'll want to add a simple if-statement around the `mov` command you are outputting - only output the `mov` if `output_spot` and `arg_spot` are not already the same.

Another more subtle and less important optimization involves whether to move before or after performing the negation. Ideally, you'll want the argument of the `neg` operand to be in a register since register access tends to be much faster than memory access. The logic could look something like this: if `output_spot` and `arg_spot` are already the same, omit the `mov`. Otherwise, if `arg_spot` is a register spot, it's probably a good idea to negate first and move second. Otherwise, emit the `mov` first and the `neg` second.

Reply:
Thanks for the really detailed explanation of how the register allocation works, Shivam. I understand the code for the IL commands a lot better now.

However, I have one question regarding the second optimization you suggested. In the case where `arg_spot` is a register, would we have to negate twice (before and after the move) so as to not modify a variable that might be stored at that register? For example, if I just negate and move, I run into issues like this:
```
...
int n = 5; 
-n; 
printf("%d", n); // -5
...
```
 Am I missing something? Thanks.

Reply:
Oops, you're totally right. Great catch :)